packages/web-components/@sfx/product/src/product.ts

```diff
@@ -0,0 +1,104 @@
+import { customElement, property, html, TemplateResult } from 'lit-element';
+
+import { Base } from '@sfx/base';
+
+@customElement('sfx-product')
+export default class Product extends Base {
+  @property({ type: Object }) product: ProductModel | any = {};
+
+  constructor() {
+    super();
+    this.updateVariant = this.updateVariant.bind(this);
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    window.addEventListener('sfx::change_product_variant', this.updateVariant);
+  }
+
+  disconnectedCallback() {
+    super.disconnectedCallback();
+    window.removeEventListener('sfx::change_product_variant', this.updateVariant);
+  }
+
+  updateVariant(e: CustomEvent) {
+    const product = e.detail;
+
+    this.product = {
+      ...this.product,
+      ...product
+    };
+  }
+
+  urlWrap(url: string, children: TemplateResult) {
+    return url ? html`<a href="${ url }">${children}</a>` : html`${children}`;
+  }
+
+  additionalInfo() {
+    const { product } = this;
+
+    const properties: ProductModel = {
+      name: '',
+      price: 0,
+      variants: { type: 'text', items: [] },
+      productUrl: '',
+      imageAlt: '',
+      imageSrc: ''
+    };
+
+    return Object.keys(product).filter(p => !(p in properties)).map(p => html`
+      <span class="sfx-product-${ p }">${ product[p] }</span>
+    `);
+  }
+
+  render() {
+    const { name, price, variants, productUrl, imageSrc, imageAlt } = this.product;
+
+    return html`
+      <style>.product-variants { padding: 0 }</style>
+      <slot name="image">
+        ${ imageSrc ? html`<img src="${ imageSrc }" alt="${ imageAlt }" />`: '' }
+      </slot>
+      <slot name="variants">
+        <ul class="product-variants">
+          ${
+            variants ?
+              variants.items.map((v: ProductVariantModel) => html`<sfx-product-variant type="${variants.type}" .variant="${v}"></sfx-product-variant>`)
+            : ''
+          }
+        </ul>
+      </slot>
+      <slot name="title">
+        ${ this.urlWrap(productUrl, html`
+          <h3>${ name }</h3>
+        `) }
+      </slot>
+      <slot name="price">
+        <p>${ price }</p>
+      </slot>
+      ${ this.additionalInfo() }
+    `;
+  }
+}
+
+export interface ProductModel {
```

Comment:
My guess why you declared the type of `this.product` as `any` is because of the possibility of additional properties? In that case, you're missing `[prop: string]: any;` in this interface. Whether you add it to `ProductModel` or you make a new interface that extends `ProductModel` to add that is up to you.

Reply:
🙏 ty needed this